pkg/gds/admin.go

```diff
@@ -1226,6 +1227,58 @@ func (s *Admin) updateVASPEndpoint(vasp *pb.VASP, commonName, endpoint, source s
 	return true, http.StatusOK, nil
 }
 
+// DeleteVASP removes a VASP and its associated certificate requests if and only if the
+// VASP verification status is in PENDING_REVIEW or earlier or ERRORED.
+func (s *Admin) DeleteVASP(c *gin.Context) {
+	var (
+		vaspID     string
+		vasp       *pb.VASP
+		certReqIDs []string
+		err        error
+	)
+
+	vaspID = c.Param("vaspID")
+
+	// Retrieve the VASP from the database
+	if vasp, err = s.db.RetrieveVASP(vaspID); err != nil {
+		log.Warn().Err(err).Msg("could not retrieve VASP from database")
+		c.JSON(http.StatusNotFound, admin.ErrorResponse("could not retrieve VASP record by ID"))
+		return
+	}
+
+	// Only allow deletions if the VASP has not been reviewed yet
+	if vasp.VerificationStatus > pb.VerificationState_PENDING_REVIEW && vasp.VerificationStatus < pb.VerificationState_ERRORED {
+		log.Warn().Str("status", vasp.VerificationStatus.String()).Msg("VASP is in invalid state for deletion")
+		c.JSON(http.StatusBadRequest, admin.ErrorResponse("cannot delete VASP in its current state"))
+		return
+	}
+
+	// Retrieve the associated certificate requests
+	if certReqIDs, err = models.GetCertReqIDs(vasp); err != nil {
+		log.Warn().Err(err).Msg("could not retrieve certificate request IDs for VASP")
+		c.JSON(http.StatusInternalServerError, admin.ErrorResponse("could not retrieve certificate requests for VASP"))
+		return
+	}
+
+	// Delete the certificate requests
+	for _, id := range certReqIDs {
+		if err = s.db.DeleteCertReq(id); err != nil {
+			log.Warn().Err(err).Str("certreq_id", id).Msg("could not delete certificate request")
+			c.JSON(http.StatusInternalServerError, admin.ErrorResponse("could not delete associated certificate request"))
+			return
+		}
+	}
```

Comment:
I thought it made sense to delete the certreqs first, because if we somehow fail to delete one of them then we still have a way to remove the VASP from the API.

Reply:
Makes sense to me!